### PR TITLE
Google auth

### PR DESCRIPTION
**Pull request by RobAWilkinson**
## Background
WIth this PR, omniauth2 is integrated and a user can login, and hit a graphql endpoint `me` to access their account data.
 
## Definition of Done
 
- [ ] User can signup with their oddball email address

### CODE REVIEW

File: app/controllers/graphql_controller.rb [hunk 1]
```diff
@@ -9,13 +9,14 @@
 # @see http://graphql-ruby.org/schema/generators#graphqlinstall
 #
 class GraphqlController < ApplicationController
+  skip_before_action :verify_authenticity_token
```
Comment (RobAWilkinson) [inline app/controllers/graphql_controller.rb]: have lots of requests coming from front_end and see no real need for csrf token checking for this

File: app/controllers/graphql_controller.rb [hunk 2]
```diff
@@ -9,13 +9,14 @@
...
   def execute
     variables = ensure_hash(graphql_params[:variables])
     query = graphql_params[:query]
     operation_name = graphql_params[:operationName]
     context = {
       # Query context goes here, for example:
-      # current_user: current_user,
+      current_user: current_user,
```
Comment (RobAWilkinson) [inline app/controllers/graphql_controller.rb]: allow for access to `current_user` in resolvers

File: app/javascript/containers/Application.js [hunk 3]
```diff
@@ -7,6 +7,32 @@ import Navbar from "../components/Navbar";
...
+      headers: {
+        "Content-Type": "application/json"
+      },
+      credentials: "same-origin"
+    }).then(resp => {
+      return resp.json()
+    }).then(data => {
+      console.log(data)
+    })
+  }
```
Comment (RobAWilkinson) [inline app/javascript/containers/Application.js]: Raw `fetch` for now to prove out while deciding with apollo or relay.

File: config/credentials.yml.enc [hunk 4]
```diff
@@ -1 +1 @@
...
\ No newline at end of file
+61M/uhq1CZFCMNw0aFPKQlnSe12NQjXQxY78vRf+/ihGwqQqOkGim68WYuTwJHj+Wi6UvZEbfXxc8lNcPbK6CLH43HOgNcXelcIe5KRUPzMuu5MqmXAw05/Z6PVrxGk7BkTbMw8NpKbXgzLXJZHl2Xoq1MRU3uOSymAP9GVRXrub95KEOJuYtHO/y7qCKBAcRkwzvTz1scDIZPpLYYiZx1eESnVGnKNOruGEU+ilSNrR4xwgIagQQKuR6HiD/UxhFIaOzJ7/2Y49JIBf6rBt6cXV2I+niWLClXmjnejWOckYDNFLQZn3lme5ScUaGYWmcIrdOu3TBjKO1l7tlXl5yxD3KUZi47tPfWLa3zsKwB0Yypq1xXj3VIhe/PwIF39e9ajLo4XxSl/5icj+z62zPVc2yQgGTjLtdtdpGLcMMIr/NESVkkaKsg1Yt/E1YLdONl19OXjLkohjJMV7kFHEfBxnoArpKJFBqskBXFQ0NEGZxKVO2b1fE8Zqf31Y3Rx+jF/+GCFIiGQkUJcD8KAob3CmlrOpxu+RMhMFoejNOqTHkrmLa/dqLxzoc6GXT1ZHMLJ9TDjInNqFYJELVU5eGv03mkT7TwPhwZkKMZ6SsxlWebNbk7BXuVjicxpJlhGK42k1N6Ntdfrfn+ZdJ1s+sTkFDT0pZYuLGENnTnqTEyoBO0AUTFyCszIZUi9mso6W0HQbTanpNdb1dBOLcsE=--mn/5yOmvj4qBY0Be--WOmbUbw97VfAZ4Bd0P0E6g==
```
Comment (RobAWilkinson) [inline config/credentials.yml.enc]: includes new `client_id` and `client_secret`

File: config/initializers/wrap_parameters.rb [hunk 5]
```diff
@@ -7,9 +7,9 @@
 
 # Enable parameter wrapping for JSON. You can disable this by
 # setting :format to an empty array.
-ActiveSupport.on_load(:action_controller) do
-  wrap_parameters format: [:json]
-end
+# ActiveSupport.on_load(:action_controller) do
+#   wrap_parameters format: [:json]
+# end
```
Comment (RobAWilkinson) [inline config/initializers/wrap_parameters.rb]: Was getting `Unpermitted parameter: :graphql` from graphql_controller. Since not building out any other API endpoints, no need for wrappind

File: spec/models/user_spec.rb [hunk 6]
```diff
@@ -6,13 +6,33 @@
...
+        email: "jane.doe@gmail.com"
+      }
+      authmock.provider = "test"
+      authmock.credentials = {
+        token: token,
+        expires_at: 1.year.from_now
+      }
+      user = User.from_omniauth(authmock)
+      expect(user).to be_a(User)
+      expect(User.count).to eq(count + 1)
```
Comment (RobAWilkinson) [inline spec/models/user_spec.rb]: test creating a user with google auth data.

File: config/initializers/wrap_parameters.rb (hunk 5 above)
Comment (hpjaj) [reply config/initializers/wrap_parameters.rb]: If you add `:graphql` to the permitted strong params in that controller, does that help?

File: app/models/user.rb [hunk 7]
```diff
@@ -9,7 +9,19 @@ class User < ApplicationRecord
                  key: Rails.application.credentials.encryption_key,
                  encode: true
 
-  validates :first_name, :last_name, presence: true
```
Comment (hpjaj) [inline app/models/user.rb]: Curious, was this to get things up and running during setup, or can we no longer validate for the presence of either of these two name attrs?

File: config/initializers/omniauth.rb [hunk 8]
```diff
@@ -0,0 +1,5 @@
+OmniAuth.config.logger = Rails.logger
+
+Rails.application.config.middleware.use OmniAuth::Builder do
+  provider :google_oauth2, Rails.application.credentials.google_auth[:client_id], Rails.application.credentials.google_auth[:client_secret], { client_options: { ssl: { ca_file: Rails.root.join("cacert.pem").to_s } } }
```
Comment (hpjaj) [inline config/initializers/omniauth.rb]: Can you pls break this apart, for readability.  e.g.

```ruby
Rails.application.config.middleware.use OmniAuth::Builder do
  provider(
    :google_oauth2, 
    Rails.application.credentials.google_auth[:client_id], 
    Rails.application.credentials.google_auth[:client_secret], 
    { 
      client_options: { 
        ssl: { ca_file: Rails.root.join("cacert.pem").to_s } 
      } 
    }
  )
end
```

File: app/models/user.rb (hunk 7 above)
Comment (RobAWilkinson) [reply app/models/user.rb]: I believe you can technically have an anonymous google account with no first name and last_name, I don't think we're going to bump against that in oddball though. I think this was just me testing, I'll add those back in

File: config/initializers/omniauth.rb (hunk 8 above)
Comment (RobAWilkinson) [inline config/initializers/omniauth.rb]:
```suggestion
Rails.application.config.middleware.use OmniAuth::Builder do
  provider(
    :google_oauth2, 
    Rails.application.credentials.google_auth[:client_id], 
    Rails.application.credentials.google_auth[:client_secret], 
    { 
      client_options: { 
        ssl: { ca_file: Rails.root.join("cacert.pem").to_s } 
      } 
    }
  )
end
```

Comment (RobAWilkinson) [reply config/initializers/omniauth.rb]: fancy new github suggestions work great.

File: config/initializers/wrap_parameters.rb (hunk 5 above)
Comment (kfrz) [reply config/initializers/wrap_parameters.rb]: IMO it's cleaner/logical to delete code rather than leave commented

File: spec/models/user_spec.rb [hunk 9]
```diff
@@ -6,13 +6,33 @@
...
       user = create :user
-
       expect(user).to be_valid
     end
 
     it { should validate_presence_of(:first_name) }
     it { should validate_presence_of(:last_name) }
     it { should validate_presence_of(:email) }
-    it { should validate_presence_of(:github_username) }
+    # it { should validate_presence_of(:github_username) }
```
Comment (kfrz) [inline spec/models/user_spec.rb]: Do we need this validation any longer? Can we just remove it, or is it saved for the future for reasons?

File: config/initializers/omniauth.rb [hunk 10]
```diff
@@ -0,0 +1,14 @@
+OmniAuth.config.logger = Rails.logger
+
+Rails.application.config.middleware.use OmniAuth::Builder do
+  provider(
+    :google_oauth2, 
+    Rails.application.credentials.google_auth[:client_id], 
+    Rails.application.credentials.google_auth[:client_secret], 
+    { 
+      client_options: { 
+        ssl: { ca_file: Rails.root.join("cacert.pem").to_s }
```
Comment (kfrz) [inline config/initializers/omniauth.rb]: Does this cert exist or will it be added in a separate PR/manual process?

Comment (RobAWilkinson) [reply config/initializers/omniauth.rb]: ha, copy pasted from a site, so weird, it was actually working and this file doesn't exist at all. Maybe a prod only type of thing that omniauth is just automagically handling?

File: spec/models/user_spec.rb (hunk 9 above)
Comment (RobAWilkinson) [reply spec/models/user_spec.rb]: is there, `validates_uniqueness_of`?

Comment (hpjaj) [reply spec/models/user_spec.rb]: No, there is not.  We'd have to hand roll it.

Comment (kfrz) [reply spec/models/user_spec.rb]: Looks like [shoulda_matchers](https://www.rubydoc.info/github/thoughtbot/shoulda-matchers/Shoulda%2FMatchers%2FActiveRecord:validate_uniqueness_of) does have a `validates_uniquness_of` matcher

Comment (hpjaj) [reply spec/models/user_spec.rb]: Nice!

Comment (RobAWilkinson) [reply spec/models/user_spec.rb]: Never used those before, is that sort of just testing that rails has that method applied? 
I guess the better way would be to create two users with the same github_username and `expect{ SecondUser.save! } to throw_error(DuplicateEntryError)`
?

Comment (kfrz) [reply spec/models/user_spec.rb]: Yes testing the behavior as you described is both more thorough and appropriate imo

Comment (hpjaj) [reply spec/models/user_spec.rb]: @RobAWilkinson - IMO, we're good to go with Thoughtbot's shoulda matchers.  You can just do:

```ruby
it { should validate_uniqueness_of(:email) }
it { should validate_uniqueness_of(:github_username) }
```

**Comment by RobAWilkinson** [comment]
This build will fail because its missing the google credentials, and the initializer needs them
